src/monogame/vulkan/MGG_Vulkan.cpp

```diff
@@ -0,0 +1,973 @@
+
+	res = vkEnumeratePhysicalDevices(system->instance, &count, gpus);
+	if (res == VK_SUCCESS)
+	{
+		for (uint32_t i = 0; i < count; ++i)
+		{
+			auto adapter = new MGG_GraphicsAdapter();
+			adapter->device = gpus[i];
+
+			vkGetPhysicalDeviceProperties(adapter->device, &adapter->properties);
+			vkGetPhysicalDeviceFeatures(adapter->device, &adapter->features);
+			vkGetPhysicalDeviceMemoryProperties(adapter->device, &adapter->memory);
+
+			system->adapters.push_back(adapter);
+		}
+	}
+
+	free((void*)gpus);
+}
+
+static MGG_System* System_Get()
+{
+	if (mgg_system != nullptr)
+		return mgg_system;
+
+	auto err = volkInitialize();
+	if (err != VK_SUCCESS)
+	{
+		printf("Failed to initialize volk!");
+		return nullptr;
+	}
+
+	VkApplicationInfo app_info = { VK_STRUCTURE_TYPE_APPLICATION_INFO };
+	app_info.pNext = nullptr;
+	app_info.applicationVersion = VK_MAKE_VERSION(1, 0, 0);
+	app_info.pApplicationName = "None";
+	app_info.pEngineName = "MonoGame";
+	app_info.engineVersion = VK_MAKE_VERSION(1, 0, 0);
+	app_info.apiVersion = VK_API_VERSION_1_0;
+
+	std::vector<const char*> instanceExtensions = { VK_KHR_SURFACE_EXTENSION_NAME };
+#ifdef _WIN32
+	instanceExtensions.push_back(VK_KHR_WIN32_SURFACE_EXTENSION_NAME);
+#endif
+	//instanceExtensions.push_back("VK_LAYER_KHRONOS_validation");
+
+	VkInstanceCreateInfo instance_create_info = { VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO };
+	instance_create_info.pApplicationInfo = &app_info;
+	instance_create_info.enabledExtensionCount = instanceExtensions.size();
+	instance_create_info.ppEnabledExtensionNames = instanceExtensions.data();
+	instance_create_info.pNext = nullptr;
+
+	VkInstance instance = VK_NULL_HANDLE;
+
+	err = vkCreateInstance(&instance_create_info, nullptr, &instance);
+	if (err != VK_SUCCESS)
+	{
+		printf("Failed to create Vulkan instance!");
+		return nullptr;
+	}
+
+	volkLoadInstance(instance);
+
+	mgg_system = new MGG_System();
+	mgg_system->instance = instance;
+
+	System_GetAdapters(mgg_system);
+
+	return mgg_system;
+}
+
+MGG_GraphicsAdapter* MGG_GraphicsAdapter_Get(mgint index)
+{
+	auto system = System_Get();
+
+	if (index < 0 || index >= system->adapters.size())
+		return nullptr;
+
+	return system->adapters[index];
+}
+
+void MGG_GraphicsAdapter_GetInfo(MGG_GraphicsAdapter* adapter, MGG_GraphicsAdaptor_Info* info)
+{
+	assert(adapter != nullptr);
+	assert(info != nullptr);
+
+	info->DeviceName = adapter->properties.deviceName;
+	info->DeviceId = adapter->properties.deviceID;
+	info->Revision = adapter->properties.driverVersion;
+	info->VendorId = adapter->properties.vendorID;
+
+	// TODO: Should we be generating a description?
+	// Is there a description somewhere for us to get?
+	static const char* description = "";
+	info->Description = (void*)description;
+
+	info->SubSystemId = 0;
+	info->MonitorHandle = 0;
+
+#ifdef _WIN32
+
+	HMONITOR primaryMonitor = MonitorFromPoint(POINT{ 0, 0 }, MONITOR_DEFAULTTOPRIMARY);
```

Comment:
This is probably all wrong and needs to be rewritten, but it works for this first pass at Vulkan.